src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1117 @@
+                surface_clone.commit();
+            }
+        });
+
+        // Assigns the toplevel role and commit
+        let xdg_toplevel = xdg_surface.get_toplevel();
+        if decoration {
+            use wayland_protocols::unstable::xdg_decoration::v1::client::zxdg_decoration_manager_v1::ZxdgDecorationManagerV1;
+
+            if let Ok(decorations) = globals
+                .instantiate_exact::<ZxdgDecorationManagerV1>(1)
+                .map_err(|e| {
+                    Error::WindowCreate(format!(
+                        "Failed creating server-side surface decoration: {:?}",
+                        e
+                    ))
+                })
+                .map_err(|e| println!("{:?}", e))
+            {
+                decorations.get_toplevel_decoration(&xdg_toplevel);
+                decorations.destroy();
+            }
+        }
+        surface.commit();
+
+        event_queue
+            .sync_roundtrip(&mut (), |_, _, _| {})
+            .map_err(|e| Error::WindowCreate(format!("Roundtrip failed: {:?}", e)))?;
+
+        // Give the surface the buffer and commit
+        surface.attach(Some(&buffer), 0, 0);
+        surface.damage(0, 0, i32::max_value(), i32::max_value());
+        surface.commit();
+
+        let xdg_config = Rc::new(RefCell::new(None));
+        {
+            let xdg_config = xdg_config.clone();
+            xdg_surface.quick_assign(move |_xdg_surface, event, _| {
+                use wayland_protocols::xdg_shell::client::xdg_surface::Event;
+
+                // Acknowledge only the last configure
+                if let Event::Configure { serial } = event {
+                    *xdg_config.borrow_mut() = Some(serial);
+                }
+            });
+        }
+
+        let cursor = wayland_cursor::load_theme(None, 16, &shm);
+        let cursor_surface = compositor.create_surface();
+
+        Ok((
+            Self {
+                _display: display,
+                wl_display,
+                _compositor: compositor,
+                _base: xdg_wm_base,
+                surface,
+                xdg_surface,
+                toplevel: xdg_toplevel,
+                _shm: shm,
+                event_queue,
+                _seat: seat,
+                xdg_config,
+                cursor,
+                cursor_surface,
+                buf_pool,
+            },
+            input,
+        ))
+    }
+
+    fn set_geometry(&self, pos: (i32, i32), size: (i32, i32)) {
+        self.xdg_surface
+            .set_window_geometry(pos.0, pos.1, size.0, size.1);
+    }
+
+    fn set_title(&self, title: &str) {
+        self.toplevel.set_title(title.to_owned());
+    }
+
+    fn set_no_resize(&self, size: (i32, i32)) {
+        self.toplevel.set_max_size(size.0, size.1);
+        self.toplevel.set_min_size(size.0, size.1);
+    }
+
+    // Sets a specific cursor style
+    fn update_cursor(&mut self, cursor: &str) {
+        let csr = self.cursor.get_cursor(cursor).unwrap();
```

Comment:
I would personally prefer to shadow `cursor`, because `csr` means absolutely nothing to me. Maybe this is what other libraries call it, though, in which case feel free to disregard.